### PR TITLE
chore(core): save initial table metadata in wal

### PR DESCRIPTION
**Pull request by ideoma**
It can be useful to get the initial state of the table to be able to restore the metadata by applying all the metadata changes already saved in WAL.

This PR also changes the mechanics of how WAL tables are renamed. Rename table is serialised to the WAL sequencer files as a structural ALTER. Wal Writers ensures that there the same table is not renamed concurrently into 2 different names. Renaming WAL table in the Name registry is done in 2 steps, first,  new name alias is created pointing to the table, then rename is serialized to WAL sequencer, and then the old table alias is removed. There is a short period of time the table can be accessed by 2 names. In case of concurrent rename, serialisation to WAL of the rename fails and the new alias is removed from the Name registry.

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 135 / 166 (81.33%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/wal/seq/TableSequencerImpl.java|0|11|00.00%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/TableSequencerAPI.java|0|4|00.00%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/TableTransactionLog.java|0|9|00.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/catalogue/WalTableListFunctionFactory.java|22|27|81.48%|
|:large_blue_circle:|io/questdb/cairo/TableWriterMetadata.java|15|16|93.75%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|62|63|98.41%|
|:large_blue_circle:|io/questdb/cairo/TableConverter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementScheduler.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/SequencerMetadata.java|18|18|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalUtils.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/CairoEngine.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/ApplyWal2TableJob.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/ColumnPurgeJob.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/AbstractTableNameRegistry.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 226 / 264 (85.61%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/wal/seq/TableTransactionLog.java|0|10|00.00%|
|:large_blue_circle:|io/questdb/cairo/TableNameRegistryRO.java|0|3|00.00%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/TableSequencerImpl.java|2|13|15.38%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/TableSequencerAPI.java|3|6|50.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/catalogue/WalTableListFunctionFactory.java|22|27|81.48%|
|:large_blue_circle:|io/questdb/cairo/CairoEngine.java|23|26|88.46%|
|:large_blue_circle:|io/questdb/cairo/TableWriterMetadata.java|15|16|93.75%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/SequencerMetadata.java|24|25|96.00%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|63|64|98.44%|
|:large_blue_circle:|io/questdb/cairo/TableConverter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/TxWriter.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementScheduler.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/SequencerMetadataService.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalWriterMetadata.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalWriter.java|15|15|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalUtils.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableNameRegistryRW.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/ApplyWal2TableJob.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/ColumnPurgeJob.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalReader.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|15|15|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/ops/AlterOperation.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cairo/AbstractTableNameRegistry.java|1|1|100.00%|

File: core/src/main/java/io/questdb/cairo/wal/WalWriter.java [hunk 1]
```diff
@@ -153,6 +154,18 @@ public WalWriter(
         }
     }
 
+
+    public long renameTable(CharSequence oldName, String newTableName) {
+        if (!Chars.equals(oldName, tableToken.getTableName())) {
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cairo/wal/WalWriter.java:159]: we need a test to assert case-insensitivity

File: core/src/main/java/io/questdb/cairo/wal/WalWriter.java [hunk 2]
```diff
@@ -1455,6 +1468,14 @@ public void addColumn(
             structureVersion++;
         }
 
+        @Override
+        public void renameTable(CharSequence oldName, CharSequence newName) {
+            if (!Chars.equals(oldName, metadata.getTableToken().getTableName())) {
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cairo/wal/WalWriter.java:1473]: same, case insensitivity

File: core/src/main/java/io/questdb/cairo/wal/seq/SequencerMetadata.java [hunk 3]
```diff
@@ -166,8 +161,14 @@ public void renameColumn(CharSequence columnName, CharSequence newName) {
         structureVersion.incrementAndGet();
     }
 
-    public void switchTo(Path path, int pathLen) {
-        assert metaMem.getFd() == -1;
+    public void renameTable(CharSequence newName) {
+        if (!Chars.equals(newName, tableToken.getTableName())) {
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cairo/wal/seq/SequencerMetadata.java:165]: case insensitivity

File: core/src/main/java/io/questdb/cairo/wal/seq/SequencerMetadataService.java [hunk 4]
```diff
@@ -49,6 +50,13 @@ public void addColumn(
         metadata.addColumn(name, type);
     }
 
+    @Override
+    public void renameTable(CharSequence oldName, CharSequence newName) {
+        assert Chars.equals(oldName, metadata.getTableToken().getTableName());
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cairo/wal/seq/SequencerMetadataService.java:55]: case insensitivity

**Review by bluestreak01 — CHANGES_REQUESTED** [review]
lets write tests to show if we deal with concurrency when table names exploit case insensitivity

File: core/src/main/java/io/questdb/cairo/CairoEngine.java [hunk 5]
```diff
@@ -724,15 +734,39 @@ public TableToken rename(
...
+                        if (newNameStr != null) {
+                            try (WalWriter walWriter = getWalWriter(tableToken)) {
+                                long seqTxn = walWriter.renameTable(tableName, newNameStr);
+                                LOG.info().$("renamed table [from='").$(tableName).$("', to='").$(newName)
+                                        .$("', wal=").$(walWriter.getWalId())
+                                        .$("', seqTxn=").$(seqTxn)
+                                        .I$();
+                                renamed = true;
+                            }
+                            TableUtils.overwriteTableNameFile(path.of(configuration.getRoot()).concat(newTableToken), memory, configuration.getFilesFacade(), newTableToken.getTableName());
```
[the remark this answers is not in the document]
Comment (ideoma) [reply core/src/main/java/io/questdb/cairo/CairoEngine.java]: thanks

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 277 / 321 (86.29%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/wal/seq/TableTransactionLog.java|0|10|00.00%|
|:large_blue_circle:|io/questdb/cairo/TableNameRegistryRO.java|0|3|00.00%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/TableSequencerImpl.java|3|14|21.43%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/TableSequencerAPI.java|3|6|50.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/catalogue/WalTableListFunctionFactory.java|22|27|81.48%|
|:large_blue_circle:|io/questdb/cairo/CairoEngine.java|69|78|88.46%|
|:large_blue_circle:|io/questdb/cairo/TableWriterMetadata.java|15|16|93.75%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/SequencerMetadata.java|28|29|96.55%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|63|64|98.44%|
|:large_blue_circle:|io/questdb/cairo/TableConverter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/TxWriter.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementScheduler.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/SequencerMetadataService.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalWriterMetadata.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalWriter.java|15|15|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalUtils.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableNameRegistryRW.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/ApplyWal2TableJob.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/ColumnPurgeJob.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalReader.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|15|15|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/ops/AlterOperation.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cairo/AbstractTableNameRegistry.java|1|1|100.00%|

File: core/src/main/java/io/questdb/cairo/wal/WalWriter.java (hunk 1 above)
Comment (ideoma) [reply core/src/main/java/io/questdb/cairo/wal/WalWriter.java]: done, converted to case insensitive check, added tests

File: core/src/main/java/io/questdb/cairo/wal/seq/SequencerMetadataService.java (hunk 4 above)
Comment (ideoma) [reply core/src/main/java/io/questdb/cairo/wal/seq/SequencerMetadataService.java]: done, converted to case insensitive check, added tests

**Review by bluestreak01 — APPROVED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 281 / 325 (86.46%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/wal/seq/TableTransactionLog.java|0|10|00.00%|
|:large_blue_circle:|io/questdb/cairo/TableNameRegistryRO.java|0|3|00.00%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/TableSequencerImpl.java|3|14|21.43%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/TableSequencerAPI.java|3|6|50.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/catalogue/WalTableListFunctionFactory.java|22|27|81.48%|
|:large_blue_circle:|io/questdb/cairo/CairoEngine.java|71|80|88.75%|
|:large_blue_circle:|io/questdb/cairo/TableWriterMetadata.java|15|16|93.75%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/SequencerMetadata.java|30|31|96.77%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|63|64|98.44%|
|:large_blue_circle:|io/questdb/cairo/TableConverter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/TxWriter.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementScheduler.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/SequencerMetadataService.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalWriterMetadata.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalWriter.java|15|15|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalUtils.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableNameRegistryRW.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/ApplyWal2TableJob.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/ColumnPurgeJob.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalReader.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|15|15|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/ops/AlterOperation.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cairo/AbstractTableNameRegistry.java|1|1|100.00%|

**Pull request merged**